### ISSUE TITLE
[Feature] Command aliases

### ISSUE THREAD

**Issue by clemyan** [yarnpkg/berry#3077]
- [x] I'd be willing to implement this feature ([contributing guide](https://yarnpkg.com/advanced/contributing))
- [x] This feature is important to have in this repository; a contrib plugin wouldn't do

**Describe the user story**

As a user of yarn, there are some commands that I run very frequently. I would like to have concise aliases for those commands so I can execute them quickly without having to remember and/or type long commands

**Describe the solution you'd like**

Add alias configuration for aliases:

```yaml
alias:
  - p: dlx @yarnpkg/pnpify --sdk
  - pv: dlx @yarnpkg/pnpify --sdk vscode
```

This configuration makes both `yarn pv` and `yarn p vscode` equivalent to `yarn dlx @yarnpkg/pnpify --sdk vscode`.

Running arbitrary shell commands (like git aliases prefixed by `!`) are out of scope due to security and complexity concerns.

**Describe the drawbacks of your solution**

A number of potentially confusing behavior:
- Due to how Clipanion command priority works, commands contributed by plugins (including built-in ones) have priority over aliases. I.e. One cannot alias `yarn add` to another command.
- If aliases have priority over scripts, then user may unknowingly execute an aliased command rather than a script they intended to run; and vice versa,

**Describe alternatives you've considered**

**Separate plugin**: Implementing alias as described as a plugin will conflict with the built-in "run" entry command. Alternatively, the plugin command can use a prefix like `yarn - p` to trigger the `p` alias. However this kind of defeats the purpose of aliases (at least for me).

**Package scripts**: Aliases can also be implemented by adding them as package scripts.

```jsonc
// package.json
{
  // ...
  "scripts": {
    "p": "yarn dlx @yarnpkg/pnpify"
  },
}
```

However, implementing aliases this way has a number of drawbacks:
- one must add all aliases they want to use (which may be many) to every project they work on
- one must manually keep global (non-project-specific) aliases in sync in all projects
- they must be made public (both in git commits and in the npm package)

On the other hand, configuring aliases in `.yarnrc.yml` means global aliases can be kept in the home configuration, which avoid all the problem above.

**Comment by arcanis** [comment]
It doesn't sound like something we'd want to do in the core

> **Separate plugin**: Implementing alias as described as a plugin will conflict with the built-in "run" entry command.

I don't think it would - conceptually, you'd just read the config before returning your plugin specs, which you'd craft dynamically (you'd create an extra command for each of your aliases). Clipanion knows how to prioritize the resolution, and a command with a static path will always be picked over a catch-all argument.

**Comment by clemyan** [comment]
> you'd just read the config before returning your plugin specs, which you'd craft dynamically (you'd create an extra command for each of your aliases)

Oh that should work. It can also just be a single command with multiple paths (?). I will try that.

Didn't consider the possibility of requiring the core outside of the command, but it makes so much sense now that you said it.

**Comment by clemyan** [comment]
Upon closer inspection, the above approach seems unviable. Plugins must return commands synchronously, Commands must configure their paths synchronously, but `Configuration.find` is async. 

Any pointers? Am I missing something?

**Comment by arcanis** [comment]
You don't have to use `Configuration.find` - you can have a simpler configuration file just for your aliases (which is likely a good idea anyway, to make it lighter).

**Comment by clemyan** [comment]
But then the plugin would need to basically reimplement the a significant portion of `Configuration.find` in a synchronous way. (E.g. the "find up" and "find home" algorithms). To me, that sounds like way too much for a simple functionality like this, especially since the core has already implemented all of those.

It would also kill perf since that logic need to run for **every** invocation of yarn, and `Configuration.find` is a significant portion of the time taken to run some of the more simple commands like `yarn --version`. I am okay with paying the cost for invocation of aliases, but not all invocations.

Asynchronous plugin instantiation sounds great, but does not solve the perf issue. Alternatively, would a hook that allows plugins to contribute entry commands be viable?

**Comment by ojkelly** [comment]
This sounds similar in a way to an issue I haven't addressed with yarn.build. In that plugin I've defined commands on `build`, `test` and `bundle`.

I've got some ideas for a command that would be on `dev`. But I haven't started on it, because I'd want to allow it to be either opted into, or remapped via a config file.

The challenge is the same here I think, essentially needing to be able to programatically name the command. If we could do that, then having a file of yarn aliases, where the aliases themselves, and the quantity of aliases are dynamic - would be possible.

I think asynchronous plugin instantiation would solve this.

**Comment by clemyan** [comment]
> I think asynchronous plugin instantiation would solve this.

Yes, but I still have my concerns about bootup performance. That is one thing I have noticed about the `interactive-tools` plugin. Since `ink` is bundled with it, parsing the `plugin-interactive-tools` binary and then instantiating `react` and `ink` takes a significant amount of time. I would think that is acceptable if that boot up cost were incurred only when using commands in the plugin, since those commands doesn't need to be particularly performant. However, the truth is that cost is incurred for every yarn invocation, even for `yarn --version`.

(To be clear, I understand why all plugins need to be instantiated for every yarn invocation. What I find unacceptable is for a plugin to incur significant boot up costs when it is not even used.)

Asynchronous instantiation helps in that multiple such plugins can instantiate in parallel. However, say I use a single external plugin, and it asynchronously instantiate by reading a file, then the boot up cost is effectively same as synchronous instantiation with `fs.readSync`.

That said, it is a "concern", because I haven't actually benchmarked this. Maybe the cost is insignificant. I will post some results here when I do.

Side note, should the loaded `Configuration` be available to plugins during instantiation? Or implement some kind of caching for `Configuration.find`? That way plugins can avoid the cost of `Configuration.find`.

**Comment by arcanis** [comment]
> However, the truth is that cost is incurred for every yarn invocation, even for `yarn --version`.

If it affects the boot time too much, such dependencies can be lazily initialized. Cf the Clipanion doc:
https://mael.dev/clipanion/docs/tips#lazy-evaluation

I've also experimented a couple of years ago with a Babel plugin to make all these imports lazy by default, but iirc it made things slightly slower (https://github.com/arcanis/babel-plugin-lazy-import). We have also switched to ESBuild since then, so the plugin wouldn't be usable out of the box anymore.

> Side note, should the loaded Configuration be available to plugins during instantiation?

The configuration isn't really available yet when the plugins are initialized, since the configuration is loaded based on what the plugins register as configuration settings. We could expose something very basic (like "the union of the raw yaml rc files"), but it's unclear how useful it would prove in practice, and would be one more thing to maintain susceptible to its own edge cases.
